File: backend/app/schemas/pipeline.py
from __future__ import annotations
from datetime import datetime
from typing import Any
from pydantic import BaseModel, Field


class PipelineNodeData(BaseModel):
    label: str = ""
    config: dict[str, Any] = Field(default_factory=dict)
    asset_id: str | None = None


class PipelineNode(BaseModel):
    id: str
    type: str
    position: dict[str, float]
    data: PipelineNodeData


class PipelineEdge(BaseModel):
    id: str
    source: str
    target: str
    sourceHandle: str
    targetHandle: str


class PipelineDefinition(BaseModel):
    nodes: list[PipelineNode]
    edges: list[PipelineEdge]
    viewport: dict[str, float] = Field(default_factory=lambda: {"x": 0, "y": 0, "zoom": 1})


class PipelineCreate(BaseModel):
    name: str
    description: str = ""
    definition: PipelineDefinition
    is_template: bool = False
    template_tags: list[str] = Field(default_factory=list)


class PipelineUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    definition: PipelineDefinition | None = None
    is_template: bool | None = None
    template_tags: list[str] | None = None


class PipelineResponse(BaseModel):
    id: str
    name: str
    description: str
    definition: dict
    is_template: bool
    template_tags: list[str]
    created_at: datetime
    updated_at: datetime
    version: int

    model_config = {"from_attributes": True}


class PipelineListResponse(BaseModel):
    items: list[PipelineResponse]
    total: int


class ValidationError(BaseModel):
    type: str
    message: str
    node_id: str | None = None
    edge_id: str | None = None
    nodes: list[str] | None = None
    source_port: str | None = None
    target_port: str | None = None
    param_name: str | None = None


class ValidationWarning(BaseModel):
    type: str
    message: str
    node_id: str | None = None


class ValidationResult(BaseModel):
    valid: bool
    errors: list[ValidationError]
    warnings: list[ValidationWarning]

File: internal/contracts/pipeline.go
package contracts

type PipelineNodeData struct {
	Label   string         `json:"label"`
	Config  map[string]any `json:"config"`
	AssetID *string        `json:"asset_id"`
}

type PipelineNode struct {
	ID       string             `json:"id"`
	Type     string             `json:"type"`
	Position map[string]float64 `json:"position"`
	Data     PipelineNodeData   `json:"data"`
}

type PipelineEdge struct {
	ID           string `json:"id"`
	Source       string `json:"source"`
	Target       string `json:"target"`
	SourceHandle string `json:"sourceHandle"`
	TargetHandle string `json:"targetHandle"`
}

type PipelineDefinition struct {
	Nodes    []PipelineNode     `json:"nodes"`
	Edges    []PipelineEdge     `json:"edges"`
	Viewport map[string]float64 `json:"viewport"`
}

type ValidationError struct {
	Type       string   `json:"type"`
	Message    string   `json:"message"`
	NodeID     *string  `json:"node_id"`
	EdgeID     *string  `json:"edge_id"`
	Nodes      []string `json:"nodes"`
	SourcePort *string  `json:"source_port"`
	TargetPort *string  `json:"target_port"`
	ParamName  *string  `json:"param_name"`
}

type ValidationWarning struct {
	Type    string  `json:"type"`
	Message string  `json:"message"`
	NodeID  *string `json:"node_id"`
}

type ValidationResult struct {
	Valid    bool                `json:"valid"`
	Errors   []ValidationError   `json:"errors"`
	Warnings []ValidationWarning `json:"warnings"`
}
